File: ArrayAndString/498. Diagonal Traverse.java
class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        int height = mat.length, width = mat[0].length;
        if(height == 0) return new int[] {};
        if(height == 1){
            return mat[0];
        }
        if(width == 1){
            int result[] = new int[height];
            for(int i = 0; i< height; i++){
                result[i] = mat[i][0];
            }
            return result;
        } //various base-cases

        int index = 0; //up and down
        int index2 = 0; //left and right
        boolean upperRight = true;
        int result[] = new int[height*width];
        int resultIndex = 0;
        while(resultIndex < result.length){
            if(upperRight){ //going upRight or downLeft?
                while(index >= 0 && index2 < width){
                    result[resultIndex] = mat[index][index2];
                    index--;
                    index2++;
                    resultIndex++;
                }
                index++;
                if(index2 >= width){
                    index2= width-1;
                    index++;
                }
            }
            else{
                while(index < height && index2 >= 0 ){
                    result[resultIndex] = mat[index][index2];
                    index++;
                    index2--;
                    resultIndex++;
                }
                index2++;
                if(index >= height){
                    index = height - 1;
                    index2++;
                }
            }
            upperRight = !upperRight;
        }
        return result;
    }
}

File: ArrayAndString/498. Diagonal Traverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        height = len(mat)
        width = len(mat[0])
        if(height == 1): return mat[0]

        result = []
        if(width ==1):
            for x in mat:
                result.append(x[0])
            return result

        index = 0 #up and down
        index2 = 0 #left and right
        upperRight = True
        
        for x in range(height * width):
            result.append(mat[index][index2])

            if(upperRight): #keep going upRight until you hit the ceiling or the right wall.
                if(index2 == width-1): #hitting the right wall
                    index+=1
                    upperRight = False
                elif(index == 0): #hitting the ceiling 
                    index2+=1 
                    upperRight = False
                else: #the normal movement
                    index-=1
                    index2+=1

            else: #downLeft until you hit the floor or left wall
                if(index == height-1): #you hit the floor 
                    index2 +=1
                    upperRight = True
                elif(index2 ==0): #hitting the left wall 
                    index+=1 
                    upperRight = True
                else: #the normal movement
                    index+=1
                    index2-=1
        return result



        